src/compositor/input/keyboard.rs

```diff
@@ -93,24 +93,45 @@ fn emit_awesome_keybindings(lua: &Lua,
     } else {
         "release"
     };
-    // TODO Should also emit by current focused client so we can
-    // do client based rules.
-    let keybindings = lua.named_registry_value::<Vec<rlua::AnyUserData>>(ROOT_KEYS_HANDLE)?;
-    for event_keysym in event.pressed_keys() {
-        for binding in &keybindings {
-            let obj: awesome::Object = binding.clone().into();
-            let key = awesome::Key::cast(obj.clone()).unwrap();
-            let keycode = key.keycode()?;
-            let keysym = key.keysym()?;
-            let modifiers = key.modifiers()?;
-            let binding_match = (keysym != 0 && keysym == event_keysym
-                                 || keycode != 0 && keycode == event.keycode())
-                                && modifiers == 0
-                                || modifiers == event_modifiers.bits();
-            if binding_match {
-                emit_object_signal(&*lua, obj, state_string.into(), event_keysym)?;
+    let mut res = Ok(());
+    // If keygrabber is set, grab key
+    // TODO check behavior when event.pressed_keys() isn't a singleton
+    if awesome::keygrabber::is_keygrabber_set(&*lua) {
+        for event_keysym in event.pressed_keys() {
+            let nres = awesome::keygrabber::call_keygrabber(&*lua, (
+                    awesome::lua::mods_to_lua(&*lua,
+                                              &awesome::lua::num_to_mods(event_modifiers))?,
+                    keysym_get_name(event_keysym),
+                    state_string.into()
+            ));
+            res = match nres {
```

Comment:
This will only report the last error, which is a bit of a shame. Can you at least log every time we set the `err` for the `res`?.

Also this line can be more clear like this:

```rust
if let Err(err) = nres {
    // log here
    res = err;
}
```

Reply:
Now that I think about it, wouldn't it be cleaner if the error was instantly returned when it happened ? Maybe it's not the default behavior we want though